## Use the following code for the questions below :

In [74]:
import numpy as np
a = np.array([4, 10, 12, 23, -2, -1, 0, 0, 0, -6, 3, -7])

## 1. How many negative numbers are there?

In [75]:
mask = (a < 0)
a [mask]

array([-2, -1, -6, -7])

In [76]:
len(a[mask])

4

### 1a. There are 4 negative numbers

## 2. How many positive numbers are there?

In [4]:
mask = (a > 0)
a [mask]

array([ 4, 10, 12, 23,  3])

### 2a. There are 5 positive numbers 

## 3. How many even positive numbers are there

In [9]:
mask = (a > 0) & (a % 2 == 0)
a [mask]

array([ 4, 10, 12])

### 3a. There are 3 even positive numbers 

## 4. If you were to add 3 to each data point, how many positive numbers would there be?

In [77]:
((a + 3) > 0).sum()


10

In [78]:
(a+3) [(a+3) > 0]

array([ 7, 13, 15, 26,  1,  2,  3,  3,  3,  6])

### 4a. There would be 10 positive numbers

## 5. If you squared each number, what would the new mean and standard deviation be?

In [17]:
np.mean(a)


3.0

In [16]:
np.mean(a ** 2)

74.0

In [18]:
np.std(a)

8.06225774829855

In [19]:
np.std(a ** 2)

144.0243035046516

### 5a. The new mean would be 74 and the new std would be 144.02

## 6. A common statistical operation on a dataset is centering. This means to adjust the data such that the mean of the data is 0. This is done by subtracting the mean from each data point. Center the data set. 

In [79]:
a

array([ 4, 10, 12, 23, -2, -1,  0,  0,  0, -6,  3, -7])

In [80]:
a.mean()

3.0

In [82]:
a_centered = a - (a.mean())
a_centered

array([  1.,   7.,   9.,  20.,  -5.,  -4.,  -3.,  -3.,  -3.,  -9.,   0.,
       -10.])

## 7. Calculate the z-score for each data point

In [83]:
(a - a.mean()) / a.std()

array([ 0.12403473,  0.86824314,  1.11631261,  2.48069469, -0.62017367,
       -0.49613894, -0.3721042 , -0.3721042 , -0.3721042 , -1.11631261,
        0.        , -1.24034735])

## 8. Copy the setup and exercise directions from More Numpy Practice into your numpy_exercises.py and add your solutions.

In [84]:
import numpy as np
a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

## Exercise 1 - Make a variable called sum_of_a to hold the sum of all the numbers in above list

In [90]:
sum(a)

55

## Exercise 2 - Make a variable named min_of_a to hold the minimum of all the numbers in the above list

In [91]:
min(a)

1

## Exercise 3 - Make a variable named max_of_a to hold the max number of all the numbers in the above list

In [92]:
max(a)

10

## Exercise 4 - Make a variable named mean_of_a to hold the average of all the numbers in the above list

In [25]:
mean_of_a = np.mean(a)
mean_of_a

5.5

## Exercise 5 - Make a variable named product_of_a to hold the product of multiplying all the numbers in the above list together

In [26]:
product_of_a = np.prod(a)
product_of_a

3628800

## Exercise 6 - Make a variable named squares_of_a. It should hold each number in a squared like [1, 4, 9, 16, 25...]

In [27]:
squares_of_a = np.square(a)
squares_of_a

array([  1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

## Exercise 7 - Make a variable named odds_in_a. It should hold only the odd numbers

In [88]:
odds_in_a = []

for i in a:
    if i % 2 == 1:
        odds_in_a.append(i)
odds_in_a

[1, 3, 5, 7, 9]

## Exercise 8 - Make a variable named evens_in_a. It should hold only the evens.


In [89]:
evens_in_a = []

for i in a:
    if i % 2 == 0:
        evens_in_a.append(i)
evens_in_a

[2, 4, 6, 8, 10]

## Setup 2: Consider what it would take to find the sum, min, max, average, sum, product, and list of squares for this list of two lists.

In [35]:
b = [
    [3, 4, 5],
    [6, 7, 8]
]

In [137]:
b = np.array(b)

## Exercise 1 - refactor the following to use numpy. Use sum_of_b as the variable. **Hint, you'll first need to make sure that the "b" variable is a numpy array**

In [38]:
sum_of_b = 0
for row in b:
    sum_of_b += sum(row)

In [138]:
np.sum(b)


33

## Exercise 2 - refactor the following to use numpy

In [40]:
min_of_b = min(b[0]) if min(b[0]) <= min(b[1]) else min(b[1])

In [135]:
np.min(b)

3

## Exercise 3 - refactor the following maximum calculation to find the answer with numpy.

In [42]:
max_of_b = max(b[0]) if max(b[0]) >= max(b[1]) else max(b[1])

In [134]:
np.max(b)

8

## Exercise 4 - refactor the following using numpy to find the mean of b

In [44]:
mean_of_b = (sum(b[0]) + sum(b[1])) / (len(b[0]) + len(b[1]))

In [133]:
np.mean(b)

5.5

## Exercise 5 - refactor the following to use numpy for calculating the product of all numbers multiplied together.

In [46]:
product_of_b = 1
for row in b:
    for number in row:
        product_of_b *= number

In [96]:
 np.prod(b)

20160

## Exercise 6 - refactor the following to use numpy to find the list of squares

In [48]:
squares_of_b = []
for row in b:
    for number in row:
        squares_of_b.append(number**2)

In [95]:
np.square(b)

array([[ 9, 16, 25],
       [36, 49, 64]])

## Exercise 7 - refactor using numpy to determine the odds_in_b

In [53]:
odds_in_b = []
for row in b:
    for number in row:
        if(number % 2 != 0):
            odds_in_b.append(number)

In [94]:
b [b % 2 == 1]

array([3, 5, 7])

## Exercise 8 - refactor the following to use numpy to filter only the even numbers

In [54]:
evens_in_b = []
for row in b:
    for number in row:
        if(number % 2 == 0):
            evens_in_b.append(number)

In [93]:
b [b % 2 == 0]

array([4, 6, 8])

## Exercise 9 - print out the shape of the array b.

In [55]:
b.shape 

(2, 3)

## Exercise 10 - transpose the array b.

In [97]:
np.transpose(b) # or b(T)

array([[3, 6],
       [4, 7],
       [5, 8]])

## Exercise 11 - reshape the array b to be a single list of 6 numbers. (1 x 6)

In [99]:
np.reshape(b, (1,6))

array([[3, 4, 5, 6, 7, 8]])

## Exercise 12 - reshape the array b to be a list of 6 lists, each containing only 1 number (6 x 1)

In [100]:
np.reshape(b, (6,1))

array([[3],
       [4],
       [5],
       [6],
       [7],
       [8]])

## Setup 3

In [102]:
c = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

In [104]:
c = np.array(c)

## Exercise 1 - Find the min, max, sum, and product of c.

In [101]:
 np.min(c)

1

In [103]:
np.max(c)

9

In [105]:
np.sum(c)

45

In [106]:
np.prod(c)


362880

## Exercise 2 - Determine the standard deviation of c.

In [107]:
np.std(c)


2.581988897471611

##  Exercise 3 - Determine the variance of c.

In [108]:
np.var(c)


6.666666666666667

## Exercise 4 - Print out the shape of the array c

In [65]:
c.shape

(3, 3)

## Exercise 5 - Transpose c and print out transposed result.

In [109]:
np.transpose(c)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

## Exercise 6 - Get the dot product of the array c with c. 

In [111]:
np.dot(c,c)

array([[ 30,  36,  42],
       [ 66,  81,  96],
       [102, 126, 150]])

## Exercise 7 - Write the code necessary to sum up the result of c times c transposed. Answer should be 261

In [112]:
(c.T * c)

array([[ 1,  8, 21],
       [ 8, 25, 48],
       [21, 48, 81]])

In [113]:
(c.T *c).sum()

261

## Write the code necessary to determine the product of c times c transposed. Answer should be 131681894400.

In [114]:
(c.T * c)

array([[ 1,  8, 21],
       [ 8, 25, 48],
       [21, 48, 81]])

In [115]:
(c.T * c).prod()

131681894400

## Setup 4

In [116]:
d = [
    [90, 30, 45, 0, 120, 180],
    [45, -90, -30, 270, 90, 0],
    [60, 45, -45, 90, -45, 180]
]

In [117]:
d = np.array(d)

## Exercise 1 - Find the sine of all the numbers in d

In [118]:
np.sin(d)

array([[ 0.89399666, -0.98803162,  0.85090352,  0.        ,  0.58061118,
        -0.80115264],
       [ 0.85090352, -0.89399666,  0.98803162, -0.17604595,  0.89399666,
         0.        ],
       [-0.30481062,  0.85090352, -0.85090352,  0.89399666, -0.85090352,
        -0.80115264]])

## Exercise 2 - Find the cosine of all the numbers in d

In [119]:
np.cos(d)

array([[-0.44807362,  0.15425145,  0.52532199,  1.        ,  0.81418097,
        -0.59846007],
       [ 0.52532199, -0.44807362,  0.15425145,  0.98438195, -0.44807362,
         1.        ],
       [-0.95241298,  0.52532199,  0.52532199, -0.44807362,  0.52532199,
        -0.59846007]])

## Exercise 3 - Find the tangent of all the numbers in d

In [120]:
np.tan(d)

array([[-1.99520041, -6.4053312 ,  1.61977519,  0.        ,  0.71312301,
         1.33869021],
       [ 1.61977519,  1.99520041,  6.4053312 , -0.17883906, -1.99520041,
         0.        ],
       [ 0.32004039,  1.61977519, -1.61977519, -1.99520041, -1.61977519,
         1.33869021]])

## Exercise 4 - Find all the negative numbers in d

In [122]:
d [d < 0]

array([-90, -30, -45, -45])

## Exercise 5 - Find all the positive numbers in d

In [123]:
d [d > 0]

array([ 90,  30,  45, 120, 180,  45, 270,  90,  60,  45,  90, 180])

## Exercise 6 - Return an array of only the unique numbers in d.

In [124]:
np.unique(d)

array([-90, -45, -30,   0,  30,  45,  60,  90, 120, 180, 270])

## Exercise 7 - Determine how many unique numbers there are in d.

In [125]:
len(np.unique(d))

11

## Exercise 8 - Print out the shape of d.

In [127]:
d.shape

(3, 6)

## Exercise 9 - Transpose and then print out the shape of d.

In [129]:
np.transpose(d)

array([[ 90,  45,  60],
       [ 30, -90,  45],
       [ 45, -30, -45],
       [  0, 270,  90],
       [120,  90, -45],
       [180,   0, 180]])

In [132]:
d.T.shape

(6, 3)

## Exercise 10 - Reshape d into an array of 9 x 2

In [131]:
np.reshape(d, (9,2))

array([[ 90,  30],
       [ 45,   0],
       [120, 180],
       [ 45, -90],
       [-30, 270],
       [ 90,   0],
       [ 60,  45],
       [-45,  90],
       [-45, 180]])